src/basic/log.c

```diff
@@ -1310,6 +1339,10 @@ void log_parse_environment_variables(void) {
         e = getenv("SYSTEMD_LOG_RATELIMIT_KMSG");
         if (e && log_set_ratelimit_kmsg_from_string(e) < 0)
                 log_warning("Failed to parse log ratelimit kmsg boolean '%s'. Ignoring.", e);
+
+        e = getenv("SYSTEMD_LOG_ASSERT_RETURN_IS_CRITICAL");
```

Comment:
I feel a bit uneasy about this switch given that it can start crashing systemd left and right :-) Another option would be a compile-time option that could be passed by the CI. (at least that's what NetworkManager did as far as I can remember)

Reply:
Looks like NM flipped it at runtime to judge from https://github.com/systemd/systemd/pull/22406#issuecomment-1029391091. I still think it should be a compile-time switch though.

Reply:
Yeah, I was planning to introduce a build time option and explicitly enable it in the CI, since this has a very high potential of screwing up the whole system.

Reply:
I think it would actually be nice if it was possible to control it at runtime too because I can imagine scenarios where it would be useful to crash resolved (or any other program linked against libsystemd) without touching networkd but that runtime switch should be hidden behind a compile time switch probably :-)

Reply:
Looks like `BUILD_MODE_DEVELOPER` should suffice here. It's already used to turn on additional assertions and it's on by default on the CI (except for Ubuntu CI and Packit).

Reply:
The environment variable and kernel option are dropped. Now, the flag is unconditionally enabled in tests and fuzzers, and enabled in all components when build with `-Dmode=developer` (default).

Reply:
I think it would make sense to keep the environment variable and turn it off by default to make it easier to test separate things separately. If I tested a hypothetical NetworkManager I probably wouldn't want `systemd` to go down along with VMs/containers. The systemd CI can always turn it on everywhere. Then again it can't always control external code linked against libsystemd so it could be useful to turn it off selectively.

Reply:
Why not to use release mode build when you want to test other software, e.g. NetworkManager?

Reply:
Because `assert_return` isn't supposed to crash in release mode but I'd want it to crash on the CI. For example here's what NM did according to https://github.com/systemd/systemd/pull/22406#issuecomment-1029403606
> NM redefines assert_return() to behave like g_return_if_fail()... which under normal conditions only prints a message.
But in NM-CI we run with G_DEBUG=fatal-warnings, which turns this into an abort().

(NM replaced those systemd parts so it's a hypothetical example)

Reply:
I took a closer look at NM and it seems the dhcpv6 client is still there so maybe it isn't a hypothetical example after all :-) Either way systemd crashes aren't that unusual to me so I don't feel strongly about this. It's just that I'm not sure what is going to happen with this option enabled everywhere if some external code linked against libsystemd starts crashing on CentOS CI once it pulls the latest version of Arch or something like that. I'll leave it to @mrc0mmand to decide :-)

Reply:
> I don't understand. We have hundreds of command-line switches which will cause the system to fail if used incorrectly

It's kind of hard to set them all to get almost all the systemd components to crash at once.

> not as bad as many others

Dunno. I turned this on globally at some point and rolled it back because even in the environment prepared to deal with all sorts of systemd crashes where PID1, journald, resolved, networkd and so on go down on a regular basis it was unbearable. I guess if weird packets are never sent or other weird things like spawning a gazillion of weird jobs are done it can be somewhat tolerated probably. Either way I think it's a debug feature and it should be available in debug builds only by analogy with the other additional assertions.

Reply:
That being said it it was in release mode it would make it easier to set it outside of, say, various CIs but I'm not sure who would need it because to, for example, instrument systemd it would still be necessary to compile systemd separately. All in all I'm for the runtime option available in debug builds only but I don't feel strongly about this.